internal/stacks/stacks.go

```diff
@@ -0,0 +1,141 @@
+	if err != nil {
+		return nil, errors.WrapIff(err, "failed to checkout commit %s", parentHead)
+	}
+	cu.Add(resetCheckout)
+
+	// Find the merge base.
+	// If our history looks like:
+	//     main: X -> Y -> Z
+	//     pr1:  X -> 1a -> 1b
+	// then the merge base is X.
+	// TODO:
+	//     This assumes that we're creating this stack branch from the current
+	//     branch which is directly based on the base branch. We'll need to try
+	//	   to read branch metadata to figure out the true parent branch if we're
+	//     creating a branch with depth >= 3.
+	mergeBase, err := repo.Git("merge-base", repo.DefaultBranch(), "HEAD")
+	if err != nil {
+		return nil, errors.WrapIff(
+			err,
+			"failed to determine merge base for commit %s and %s",
+			git.ShortSha(parentHead), repo.DefaultBranch(),
+		)
+	}
+	if mergeBase == "" {
+		return nil, errors.New("merge base is empty")
+	}
+	logrus.WithField("commit", git.ShortSha(mergeBase)).Debug("determined merge base")
+
+	// Create a squash commit by taking the tree of the HEAD commit and
+	// parenting the commit on top of the merge base.
+	// Using the sample history above, the squash commit would look like:
+	//     av-base/<name>: X -> 1S
+	// where 1S is the squash of 1a and 1b.
+	// TODO:
+	//     Make this commit message better. Should probably contain commit
+	//     titles of each of the constituent commits.
+	squashCommit, err := repo.Git("commit-tree", "HEAD^{tree}", "-p", mergeBase, "-m", "[[ av synthetic squash commit ]]")
+	if err != nil {
+		return nil, errors.WrapIff(
+			err,
+			"failed to create squash commit for branch",
+		)
+	}
+	logrus.WithField("sha", git.ShortSha(squashCommit)).Debug("created synthetic squash commit")
+
+	// Save the squash commit as a new branch.
+	baseRefName := "refs/av/stack-base/" + opts.Name
+	if err := repo.UpdateRef(&git.UpdateRef{Ref: baseRefName, New: squashCommit}); err != nil {
+		return nil, err
+	}
+	logrus.WithField("ref", baseRefName).Debug("created base ref")
+
+	// Create the new branch and check it out
+	headRefName := "refs/heads/" + opts.Name
+	if err := repo.UpdateRef(&git.UpdateRef{
+		Ref: headRefName,
+		New: squashCommit,
+		Old: git.Missing,
+	}); err != nil {
+		logrus.WithError(err).Debug("failed to create new branch, it probably already exists")
+		return nil, errors.Errorf("cannot create branch %s (does it already exist?)", opts.Name)
+	}
+	// Note: this can't include refs/head/... because that enters a detached
+	// HEAD state.
+	if _, err := repo.Git("checkout", opts.Name); err != nil {
+		return nil, errors.WrapIff(err, "failed to checkout new branch %s", opts.Name)
+	}
+
+	// Finally, write the metadata and return the branch.
+	next := &Branch{
+		Name:   opts.Name,
+		Parent: repo.DefaultBranch(),
+	}
+	if err := writeStackMetadata(repo, next); err != nil {
+		return nil, err
+	}
+	cu.Cancel()
+	return next, nil
+}
+
+func writeStackMetadata(repo *git.Repo, s *Branch) error {
+	refName := "refs/av/stack-metadata/" + s.Name
```

Comment:
where would we use this?

Reply:
Currently, we only support creating a stack-branch that's twice-removed from `main`.

So if we have `main -> feature1` (`feature1` is just a "normal" branch), we can create `feature1 -> feature2` by squashing all the commits from `feature2`. However, to do it for `feature2 -> feature3`, we need to be able to figure out that `feature2` is stacked on top of `feature1` -- which it can't right now.

I also think that reading all the stackmetadata will be how we construct the tree (i.e., list all the stackmetadata refs and construct a tree out of them).